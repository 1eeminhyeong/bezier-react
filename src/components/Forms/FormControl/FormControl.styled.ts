/* Internal dependencies */
import { styled } from 'Foundation'
import { InterpolationProps } from 'Types/Foundation'

export const Wrapper = styled.div<InterpolationProps>`
  position: relative;
  ${({ interpolation }) => interpolation}
`

export const LabelWrapper = styled.div`
  padding: 0 2px;
  margin-bottom: 4px;
`

export const HelperTextWrapper = styled.div`
  padding: 0 2px;
  margin-top: 4px;
`
