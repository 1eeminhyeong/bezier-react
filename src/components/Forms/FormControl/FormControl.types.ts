/* Internal dependencies */
import type { BezierComponentProps, ChildrenProps, IdentifierProps } from 'Types/ComponentProps'
import type { FormComponentProps } from 'Components/Forms/Form.types'

interface FormControlOptions {
  labelPosition?: 'top' | 'left'
}

interface FormControlContextCommonValue extends IdentifierProps {
  Wrapper: React.FunctionComponent
}

type PropsGetter<ExtraReturnType = {}> = <Props = {}>(props: Props)
  => Props & FormControlContextCommonValue & ExtraReturnType

export type LabelPropsGetter = PropsGetter

export type FieldPropsGetter = PropsGetter<{
  'aria-describedby'?: string
}>

export type HelperTextPropsGetter = PropsGetter<{
  setHasHelperText: React.Dispatch<React.SetStateAction<boolean>>
}>

export interface FormContextValue extends FormComponentProps {
  id: string
  labelId: string
  helperTextId: string
  getLabelProps: LabelPropsGetter
  getFieldProps: FieldPropsGetter
  getHelperTextProps: HelperTextPropsGetter
}

export default interface FormControlProps extends
  BezierComponentProps,
  ChildrenProps,
  Omit<FormComponentProps, 'aria-describedby'>,
  FormControlOptions {}
