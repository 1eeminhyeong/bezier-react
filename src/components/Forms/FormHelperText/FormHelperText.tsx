/* External dependencies */
import React, { forwardRef, useEffect } from 'react'
import { isEmpty } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import useFormControlContext from 'Components/Forms/useFormControlContext'
import type FormHelperTextProps from './FormHelperText.types'
import * as Styled from './FormHelperText.styled'

export const FORM_HELPER_TEXT_TEST_ID = 'bezier-react-form-helper-text'

function FormHelperText({
  testId = FORM_HELPER_TEXT_TEST_ID,
  as = 'p',
  typo = Typography.Size13,
  children,
  ...rest
}: FormHelperTextProps,
forwardedRef: React.Ref<HTMLParamElement>,
) {
  const contextValue = useFormControlContext()

  const {
    hasError,
    setHasHelperText,
    Wrapper,
    ...ownProps
  } = contextValue?.getHelperTextProps(rest) ?? {
    ...rest,
    setHasHelperText: undefined,
    Wrapper: React.Fragment,
  }

  const shouldRendered = !isEmpty(children)

  useEffect(() => {
    setHasHelperText?.(shouldRendered)
  }, [
    shouldRendered,
    setHasHelperText,
  ])

  useEffect(() => function cleanUp() {
    setHasHelperText?.(false)
  }, [setHasHelperText])

  if (!shouldRendered) { return null }

  return (
    <Wrapper>
      <Styled.HelperText
        {...ownProps}
        aria-live={hasError ? 'polite' : undefined}
        color={hasError ? 'bgtxt-orange-normal' : 'txt-black-dark'}
        testId={testId}
        ref={forwardedRef}
        forwardedAs={as}
        typo={typo}
      >
        { children }
      </Styled.HelperText>
    </Wrapper>
  )
}

export default forwardRef(FormHelperText)
