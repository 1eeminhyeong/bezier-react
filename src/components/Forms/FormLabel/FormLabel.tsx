/* External dependencies */
import React, { forwardRef, useMemo } from 'react'
import { isEmpty } from 'lodash-es'

/* Internal dependencies */
import { Typography } from 'Foundation'
import { IconSize } from 'Components/Icon'
import type FormLabelProps from './FormLabel.types'
import * as Styled from './FormLabel.styled'

export const FORM_LABEL_TEST_ID = 'bezier-react-form-label'
export const FORM_LABEL_HELP_TEST_ID = 'bezier-react-form-label-help'

function FormLabel({
  testId = FORM_LABEL_TEST_ID,
  help,
  as = 'label',
  bold = true,
  typo = Typography.Size13,
  marginBottom,
  style,
  children,
  ...rest
}: FormLabelProps,
forwardedRef: React.Ref<HTMLLabelElement>,
) {
  const LabelComponent = useMemo(() => (
    <Styled.Label
      {...rest}
      ref={forwardedRef}
      testId={testId}
      forwardedAs={as}
      bold={bold}
      typo={typo}
    >
      { children }
    </Styled.Label>
  ), [
    rest,
    as,
    typo,
    bold,
    testId,
    children,
    forwardedRef,
  ])

  if (isEmpty(children)) { return null }

  return isEmpty(help)
    ? (
      <>
        { LabelComponent }
      </>
    )
    : (
      <Styled.Box>
        { LabelComponent }
        <Styled.Tooltip content={help}>
          <Styled.HelpIcon
            testId={FORM_LABEL_HELP_TEST_ID}
            name="help-filled"
            size={IconSize.XS}
            color="txt-black-dark"
          />
        </Styled.Tooltip>
      </Styled.Box>
    )
}

export default forwardRef(FormLabel)
