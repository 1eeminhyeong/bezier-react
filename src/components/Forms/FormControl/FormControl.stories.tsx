/* External dependencies */
import React from 'react'
import base from 'paths.macro'
import type { Story, Meta } from '@storybook/react'

/* Internal dependencies */
import { getTitle } from 'Utils/storyUtils'
import { SegmentedControl } from 'Components/Forms/SegmentedControl'
import { Radio } from 'Components/Forms/Radio'
import { Checkbox } from 'Components/Forms/Checkbox'
import { Text } from 'Components/Text'
import { FormLabel } from 'Components/Forms/FormLabel'
import { TextField } from 'Components/Forms/Inputs/TextField'
import { TextArea } from 'Components/Forms/Inputs/TextArea'
import { Select } from 'Components/Forms/Inputs/Select'
import { FormHelperText } from 'Components/Forms/FormHelperText'
import { FormErrorMessage } from 'Components/Forms/FormErrorMessage'
import FormControl from './FormControl'
import FormControlProps from './FormControl.types'

export default {
  title: getTitle(base),
  component: FormControl,
} as Meta

const Template: Story<FormControlProps> = ({ hasError, ...args }) => (
  <FormControl
    {...args}
    style={{ width: 400 }}
    hasError={hasError}
  >
    <FormLabel help="Lorem Ipsum">Label</FormLabel>
    <TextField placeholder="Placeholder" />
    <FormHelperText>Description</FormHelperText>
    <FormErrorMessage>Error!</FormErrorMessage>
  </FormControl>
)

export const Primary: Story<FormControlProps> = Template.bind({})
Primary.args = {
  id: 'form',
  labelPosition: 'top',
  hasError: false,
  disabled: false,
  readOnly: false,
}

const WithMultiFormTemplate: Story<FormControlProps> = ({ hasError, ...args }) => (
  <div style={{
    display: 'flex',
    flexDirection: 'column',
    gap: 16,
    width: 500,
  }}
  >
    <FormControl
      {...args}
      hasError={hasError}
    >
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <TextField placeholder="Placeholder" />
      <FormHelperText>{ hasError ? 'Error!' : 'Description' }</FormHelperText>
    </FormControl>
    <FormControl
      {...args}
      id={`${args.id}-1`}
      hasError={hasError}
    >
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <TextArea placeholder="Placeholder" />
      <FormHelperText>{ hasError ? 'Error!' : 'Description' }</FormHelperText>
    </FormControl>
    <FormControl
      {...args}
      id={`${args.id}-2`}
      hasError={hasError}
    >
      <FormLabel>Label</FormLabel>
      <Select />
    </FormControl>
    <FormControl
      {...args}
      id={`${args.id}-3`}
      hasError={hasError}
    >
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <SegmentedControl>
        { ['Open', 'Snoozed', 'Closed'].map((item) => (
          <Text key={item} bold>{ item }</Text>
        )) }
      </SegmentedControl>
      <FormHelperText>{ hasError ? 'Error!' : 'Description' }</FormHelperText>
    </FormControl>
    <FormControl
      {...args}
      id={`${args.id}-4`}
      hasError={hasError}
    >
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <Checkbox>Option</Checkbox>
      <FormHelperText>{ hasError ? 'Error!' : 'Description' }</FormHelperText>
    </FormControl>
    <FormControl
      {...args}
      id={`${args.id}-5`}
      hasError={hasError}
    >
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <Radio>Option</Radio>
      <FormHelperText>{ hasError ? 'Error!' : 'Description' }</FormHelperText>
    </FormControl>
  </div>
)

export const WithMultiForm: Story<FormControlProps> = WithMultiFormTemplate.bind({})
WithMultiForm.args = {
  id: 'form',
  labelPosition: 'top',
  hasError: false,
  disabled: false,
  readOnly: false,
}
