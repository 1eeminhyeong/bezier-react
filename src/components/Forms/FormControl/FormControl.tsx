/* External dependencies */
import React, { useState, useCallback, useMemo } from 'react'

/* Internal dependencies */
import useId from 'Hooks/useId'
import { omitBezierComponentProps, pickBezierComponentProps } from 'Utils/propsUtils'
import FormControlContext from './FormControlContext'
import FormControlProps, { FieldPropsGetter, LabelPropsGetter, HelperTextPropsGetter } from './FormControl.types'
import * as Styled from './FormControl.styled'

// TODO: 테스트 작성
function FormControl({
  id: idProps,
  testId,
  labelPosition = 'top',
  children,
  ...rest
}: FormControlProps) {
  const [hasHelperText, setHasHelperText] = useState(false)

  const uuid = useId()
  const id = idProps ?? `field-${uuid}`
  const labelId = `${id}-label`
  const helperTextId = `${id}-help-text`

  const bezierProps = useMemo(() => pickBezierComponentProps(rest), [rest])
  const formCommonProps = useMemo(() => omitBezierComponentProps(rest), [rest])

  const getLabelProps = useCallback<LabelPropsGetter>(ownProps => ({
    id: labelId,
    htmlFor: id,
    Wrapper: labelPosition === 'top'
      ? Styled.TopLabelWrapper
      : Styled.LeftLabelWrapper,
    ...ownProps,
  }), [
    id,
    labelId,
    labelPosition,
  ])

  const getFieldProps = useCallback<FieldPropsGetter>(ownProps => ({
    id,
    'aria-describedby': hasHelperText
      ? helperTextId
      : undefined,
    Wrapper: Styled.FieldWrapper,
    ...formCommonProps,
    ...ownProps,
  }), [
    id,
    helperTextId,
    hasHelperText,
    formCommonProps,
  ])

  const getHelperTextProps = useCallback<HelperTextPropsGetter>(ownProps => ({
    id: helperTextId,
    setHasHelperText,
    hasError: formCommonProps?.hasError,
    Wrapper: labelPosition === 'top'
      ? Styled.TopHelperTextWrapper
      : Styled.LeftHelperTextWrapper,
    ...ownProps,
  }), [
    helperTextId,
    labelPosition,
    formCommonProps,
  ])

  const rootProps = useMemo(() => ({
    role: 'group',
    ...bezierProps,
  }), [bezierProps])

  const contextValue = useMemo(() => ({
    id,
    labelId,
    helperTextId,
    getLabelProps,
    getFieldProps,
    getHelperTextProps,
    ...formCommonProps,
  }), [
    id,
    labelId,
    helperTextId,
    getLabelProps,
    getFieldProps,
    getHelperTextProps,
    formCommonProps,
  ])

  if (!children) { return null }

  const Container = {
    top: Styled.Box,
    left: Styled.Grid,
  }[labelPosition]

  return (
    <FormControlContext.Provider value={contextValue}>
      <Container
        data-testid={testId}
        {...rootProps}
      >
        { children }
      </Container>
    </FormControlContext.Provider>
  )
}

export default FormControl
