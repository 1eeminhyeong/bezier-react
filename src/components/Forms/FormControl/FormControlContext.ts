/* External dependencies */
import { createContext } from 'react'

/* Internal dependencies */
import { FormContextValue } from 'Components/Forms/FormControl'

const FormControlContext = createContext<FormContextValue | undefined>(undefined)

export default FormControlContext
