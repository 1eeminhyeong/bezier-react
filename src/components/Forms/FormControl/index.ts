import FormControl from './FormControl'
import FormControlContext from './FormControlContext'
import type FormControlProps from './FormControl.types'
import type { FormControlContextValue, FormControlContextCommonValue } from './FormControl.types'

export type {
  FormControlProps,
  FormControlContextValue,
  FormControlContextCommonValue,
}

export {
  FormControl,
  FormControlContext,
}
