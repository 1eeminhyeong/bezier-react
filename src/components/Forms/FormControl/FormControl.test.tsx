/* External dependencies */
import React from 'react'

/* Internal dependencies */
import { render } from 'Utils/testUtils'
import { FormLabel } from 'Components/Forms/FormLabel'
import { TextField } from 'Components/Forms/Inputs/TextField'
import { FormHelperText, FormErrorMessage } from 'Components/Forms/FormHelperText'
import FormControl, { FORM_CONTROL_TEST_ID } from './FormControl'
import type FormControlProps from './FormControl.types'

// TODO: 테스트 보강
describe('FormControl >', () => {
  let props: FormControlProps

  beforeEach(() => {
    props = {
      id: 'form',
      labelPosition: 'top',
      leftLabelWrapperHeight: undefined,
      hasError: false,
      disabled: false,
      readOnly: false,
    }
  })

  const renderComponent = ({ children, ...rest }: FormControlProps) => render(
    <FormControl {...props} {...rest}>
      { children }
    </FormControl>,
  )

  const defaultChildren = (
    <>
      <FormLabel help="Lorem Ipsum">Label</FormLabel>
      <TextField placeholder="Placeholder" />
      <FormHelperText>Description</FormHelperText>
      <FormErrorMessage>Error!</FormErrorMessage>
    </>
  )

  it('Snapshot >', () => {
    const { getByTestId } = renderComponent({
      children: defaultChildren,
    })

    const rendered = getByTestId(FORM_CONTROL_TEST_ID)

    expect(rendered).toMatchSnapshot()
  })
})
