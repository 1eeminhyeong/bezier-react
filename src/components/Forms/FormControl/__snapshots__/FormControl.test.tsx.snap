// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormControl > Snapshot > 1`] = `
.c7 {
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  margin: 0px 0px 0px 0px;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c9 {
  width: 100%;
  height: 100%;
  padding: 0;
  cursor: auto;
  background-color: transparent;
  border: none;
  outline: none;
  font-size: 14px;
  line-height: 18px;
  color: #000000D9;
}

.c9::-webkit-input-placeholder {
  color: #00000066;
}

.c9::-moz-placeholder {
  color: #00000066;
}

.c9:-ms-input-placeholder {
  color: #00000066;
}

.c9::placeholder {
  color: #00000066;
}

.c8 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 36px;
  padding: 0 12px;
  background-color: #FCFCFC;
  overflow: hidden;
  border-radius: 8px;
  box-shadow: 0 1px 2px #0000000D, inset 0 0 0 1px #00000026;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: border-color,box-shadow;
  transition-property: border-color,box-shadow;
}

.c0 {
  position: relative;
}

.c1 {
  padding: 0 2px;
  margin-bottom: 4px;
}

.c10 {
  padding: 0 2px;
  margin-top: 4px;
}

.c3 {
  font-size: 13px;
  line-height: 18px;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: bold;
  color: #000000D9;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c11 {
  font-size: 13px;
  line-height: 18px;
  margin: 0px 0px 0px 0px;
  font-style: normal;
  font-weight: normal;
  color: #00000066;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: color;
  transition-property: color;
}

.c5 {
  position: relative;
}

.c4 {
  display: block;
  text-align: left;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-left: 6px;
}

.c6:hover > .{
  color: #000000D9;
}

.c12 {
  display: block;
  text-align: left;
}

<div
  class="c0"
  data-testid="bezier-react-form-control"
>
  <div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <label
        class="c3 c4"
        color="txt-black-darkest"
        data-testid="bezier-react-form-label"
        for="form"
        id="form-label"
      >
        Label
      </label>
      <div
        class="c5 c6"
        data-testid="bezier-react-tooltip"
      >
        <svg
          class="c7 "
          color="txt-black-dark"
          data-testid="bezier-react-form-label-help"
          fill="none"
          foundation="[object Object]"
          height="16"
          marginbottom="0"
          marginleft="0"
          marginright="0"
          margintop="0"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12 2c5.523 0 10 4.477 10 10s-4.477 10-10 10S2 17.523 2 12 6.477 2 12 2zm.83 12.541h-1.9v-.708c0-1.156.625-2.24 1.672-2.9.867-.546 1.337-.939 1.337-1.447 0-1.014-1-1.561-1.94-1.561-1.035 0-1.942.73-1.942 1.56h-1.9c0-1.875 1.758-3.46 3.841-3.46 2.154 0 3.842 1.52 3.842 3.46 0 1.655-1.393 2.533-2.225 3.056-.236.15-.784.573-.784 1.292v.708zm-.889 3.631a1.244 1.244 0 110-2.487 1.244 1.244 0 010 2.487z"
            fill="currentColor"
            fill-rule="evenodd"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    class="c8"
    data-testid="bezier-react-text-input"
    size="36"
  >
    <input
      aria-describedby="form-help-text"
      autocomplete="off"
      class="c9"
      id="form"
      placeholder="Placeholder"
      size="36"
      value=""
    />
  </div>
  <div
    class="c0 c10"
  >
    <p
      class="c11 c12"
      color="txt-black-dark"
      data-testid="bezier-react-form-helper-text"
      id="form-help-text"
    >
      Description
    </p>
  </div>
</div>
`;
